### PR TITLE
[macOS] Fix xtro failures introduced by the xcode9 branch

### PR DESCRIPTION
**Pull request by timrisi**

### CODE REVIEW

**Comment by chamons** [comment]
This will need to go into 15.5.

**Comment by timrisi** [comment]
Checked the API diff and it only shows changes for the 2 removed mac APIs

File: tests/xtro-sharpie/osx.ignore [hunk 1]
```diff
@@ -445,3 +445,11 @@
 ## PDFKit
 # The following enums are bound as PdfFoo, lowercase Pdf prefix
 !missing-enum! PDFPrintScalingMode not bound
+
+# Xcode 9
+## xtro simultaniously says missing and existing binding is unknown
+!missing-enum! NSURLSessionDelayedRequestDisposition not bound
```
[the remark this answers is not in the document]
Comment (spouliot) [reply tests/xtro-sharpie/osx.ignore]: casing `URL` vs `Url`
Add a mapping in `Helpers.cs`
and remove it from the files

File: tests/xtro-sharpie/osx.ignore [hunk 2]
```diff
@@ -445,3 +445,11 @@
...
+!unknown-native-enum! NSUrlSessionDelayedRequestDisposition bound
+
+## Exists in the headers and documentation for macOS 10.13
+!unknown-type! WKSnapshotConfiguration bound
```
Comment (spouliot) [inline tests/xtro-sharpie/osx.ignore]: double check, it's possible that the specific header file is *not* included in the _umbrella_ header

File: tests/xtro-sharpie/osx.pending [hunk 3]
```diff
@@ -344,6 +344,21 @@
...
+!incorrect-protocol-member! NSAccessibility::setAccessibilityCustomRotors: is REQUIRED and should be abstract
+!incorrect-protocol-member! CIImageProcessorInput::surface is REQUIRED and should be abstract
+!incorrect-protocol-member! CIImageProcessorOutput::surface is REQUIRED and should be abstract
+!incorrect-protocol-member! INSpeakable::alternativeSpeakableMatches is REQUIRED and should be abstract
+!incorrect-protocol-member! INSpeakable::vocabularyIdentifier is REQUIRED and should be abstract
+!wrong-enum-size! NSSpellingState managed 4 vs native 8
+!extra-protocol-member! unexpected selector MTLCommandBuffer::GPUEndTime found
+!extra-protocol-member! unexpected selector MTLCommandBuffer::GPUStartTime found
+!extra-protocol-member! unexpected selector MTLCommandBuffer::kernelEndTime found
+!extra-protocol-member! unexpected selector MTLCommandBuffer::kernelStartTime found
```
Comment (spouliot) [inline tests/xtro-sharpie/osx.pending]: the above `extra-` needs justification - is it available ? (or a past mistake)

File: tests/xtro-sharpie/osx.pending [hunk 4]
```diff
@@ -344,6 +344,21 @@
...
+!incorrect-protocol-member! NSAccessibility::accessibilityCustomActions is REQUIRED and should be abstract
+!incorrect-protocol-member! NSAccessibility::accessibilityCustomRotors is REQUIRED and should be abstract
+!incorrect-protocol-member! NSAccessibility::setAccessibilityChildrenInNavigationOrder: is REQUIRED and should be bstract
+!incorrect-protocol-member! NSAccessibility::setAccessibilityCustomActions: is REQUIRED and should be abstract
+!incorrect-protocol-member! NSAccessibility::setAccessibilityCustomRotors: is REQUIRED and should be abstract
+!incorrect-protocol-member! CIImageProcessorInput::surface is REQUIRED and should be abstract
+!incorrect-protocol-member! CIImageProcessorOutput::surface is REQUIRED and should be abstract
+!incorrect-protocol-member! INSpeakable::alternativeSpeakableMatches is REQUIRED and should be abstract
+!incorrect-protocol-member! INSpeakable::vocabularyIdentifier is REQUIRED and should be abstract
+!wrong-enum-size! NSSpellingState managed 4 vs native 8
```
Comment (spouliot) [inline tests/xtro-sharpie/osx.pending]: It's a breaking change (if not new) but it's also a mistake.
API that use this enum must be verified to make sure they are correct.

File: tests/xtro-sharpie/osx.pending (hunk 3 above)
Comment (timrisi) [reply tests/xtro-sharpie/osx.pending]: This is a past mistake.  They were added as available for 10.12.4 a while back, but are only actually available for iOS and tvOS.  I added #if !XAMCORE_4_0 || !MONOMAC to remove them from mac for XAMCORE_4_0

File: tests/xtro-sharpie/osx.pending (hunk 4 above)
Comment (timrisi) [reply tests/xtro-sharpie/osx.pending]: This was previously used as a convenience enum in our bindings from an untyped header enum:
// Untyped enum (NSAttributedString.h). Only used as a convience enum in our API.

They made it an explicit NSInteger enum in 10.13

File: src/AppKit/XEnums.cs [hunk 5]
```diff
@@ -91,7 +91,13 @@ public enum NSCharacterCollection : nuint_compat_int {
 
 	// Untyped enum (NSAttributedString.h). Only used as a convience enum in our API.
 	[Flags]
-	public enum NSSpellingState : int {
+	public enum NSSpellingState :
+#if XAMCORE_4_0
+		nint
```
Comment (spouliot) [inline src/AppKit/XEnums.cs]: that requires `[Native]` on the type

File: tests/xtro-sharpie/osx.ignore [hunk 6]
```diff
@@ -444,4 +444,4 @@
 
 ## PDFKit
 # The following enums are bound as PdfFoo, lowercase Pdf prefix
-!missing-enum! PDFPrintScalingMode not bound
+!missing-enum! PDFPrintScalingMode not bound
```
Comment (spouliot) [inline tests/xtro-sharpie/osx.ignore]: white noise, the file has no change and should not be in the PR

File: tests/xtro-sharpie/osx.pending (hunk 4 above)
Comment (spouliot) [reply tests/xtro-sharpie/osx.pending]: ok, transform that into a comment in the .pending file

File: tests/xtro-sharpie/osx.pending (hunk 3 above)
Comment (spouliot) [reply tests/xtro-sharpie/osx.pending]: also add a comment about that

**Comment by chamons** [comment]
@spouliot I used some diff/comm/filtering trickery to ask the question "what xtro lines in osx.results are newly introduced by merging xcode9 and we should look at most closely"

That gave this list:

https://gist.github.com/chamons/33278d78eab20814dc5a63beff1823e1

which I then compared to the delta added by this file:

https://github.com/xamarin/xamarin-macios/commit/a6d978118ea01d5db53304cb17ecf4fa11c550b4#diff-12af96a98b68f09bc3eecb1e4acb65ec

since those were issues forever that Rolf's change "added" by fixing the check.

This PR fixes that subset, which seemed like a good first step.

**Comment by spouliot** [comment]
@chamons I'm fine with any subset, let's just make the title and commit message reflect this :)

**Comment by chamons** [comment]
@spouliot I think "[macOS] Fix xtro failures introduced by the xcode9 branch" is pretty decent. Thoughts/Suggestions?

**Comment by spouliot** [comment]
@chamons when I read the title 

> Fix xtro failures introduced by the xcode9 branch

I expected they were *all* fixed, but your comment

> This PR fixes that subset, which seemed like a good first step.

is right - it's a subset. Just prepend "Partial" to the title so no one thinks it's complete yet (harder to review if you think it's complete and see missing stuff).

**Comment by timrisi** [comment]
It fixes all of the failures *introduced* by the xcode 9 merge.  It doesn't fix the ones that were already there before that.

File: src/metal.cs [hunk 7]
```diff
@@ -314,6 +315,7 @@ partial interface MTLCommandBuffer {
 		[iOS (10,3)][TV (10,2)][Mac (10,12,4, onlyOn64 : true)]
 		[Export ("GPUEndTime")]
 		double /* CFTimeInterval */ GpuEndTime { get; }
+#endif
```
Comment (VincentDondain) [inline src/metal.cs]: Minor: better add `// !XAMCORE_4_0 || !MONOMAC` next to the `#endif`.

File: src/metal.cs [hunk 8]
```diff
@@ -287,6 +287,7 @@ partial interface MTLCommandBuffer {
 		[Export ("renderCommandEncoderWithDescriptor:")]
 		IMTLRenderCommandEncoder CreateRenderCommandEncoder (MTLRenderPassDescriptor renderPassDescriptor);
 
+#if !XAMCORE_4_0 || !MONOMAC
```
[the remark this answers is not in the document]
Comment (VincentDondain) [reply src/metal.cs]: Maybe a comment explaining why it's there would help (;

**Pull request merged**